src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs

```diff
@@ -86,43 +124,42 @@ private void AddOrUpdateMetricsCache(string key, long value)
                 });
         }
 
-        public IReadOnlyDictionary<string, long>? GetHostMetricsOrNull()
+        public IReadOnlyDictionary<string, long> GetHostMetricsOrNull()
         {
             if (!HasMetrics())
             {
                 return null;
             }
 
-            try
+            var functionActivityStatusProvider = _serviceProvider.GetScriptHostServiceOrNull<IFunctionActivityStatusProvider>();
+            if (functionActivityStatusProvider is not null)
             {
-                var functionActivityStatusProvider = _serviceProvider.GetScriptHostServiceOrNull<IFunctionActivityStatusProvider>();
-                if (functionActivityStatusProvider is not null)
-                {
-                    var functionActivityStatus = functionActivityStatusProvider.GetStatus();
-                    AddOrUpdateMetricsCache(HostMetrics.ActiveInvocationCount, functionActivityStatus.OutstandingInvocations);
-                }
-
-                return new Dictionary<string, long>(_metricsCache);
+                var functionActivityStatus = functionActivityStatusProvider.GetStatus();
+                AddOrUpdateMetricsCache(HostMetrics.ActiveInvocationCount, functionActivityStatus.OutstandingInvocations);
             }
-            finally
+
+            IReadOnlyDictionary<string, long> metrics;
+
+            lock (_lock)
             {
-                PurgeMetricsCache();
+                metrics = new Dictionary<string, long>(_metricsCache);
```

Comment:
Initially defined as concurrent, but here not concurrent

Reply:
I just figured that we only need a readonly copy of the dictionary for writing to the file, the metrics publisher does not need this to be concurrent?